timemachine/cpp/src/nonbonded_interaction_group.cu

```diff
@@ -0,0 +1,480 @@
+            gpuErrchk(cudaMemcpyAsync(
+                d_perm_ + NC_, d_row_atom_idxs_, NR_ * sizeof(*d_row_atom_idxs_), cudaMemcpyDeviceToDevice, stream));
+        }
+
+        // compute new coordinates, new lambda_idxs, new_plane_idxs
+        k_permute<<<dimGrid, tpb, 0, stream>>>(N_, d_perm_, d_x, d_sorted_x_);
+        gpuErrchk(cudaPeekAtLastError());
+
+        nblist_.build_nblist_device(
+            NC_, NR_, d_sorted_x_, d_sorted_x_ + 3 * NC_, d_box, cutoff_ + nblist_padding_, stream);
+        gpuErrchk(cudaMemcpyAsync(
+            p_ixn_count_, nblist_.get_ixn_count(), 1 * sizeof(*p_ixn_count_), cudaMemcpyDeviceToHost, stream));
+
+        std::vector<double> h_box(9);
+        gpuErrchk(cudaMemcpyAsync(&h_box[0], d_box, 3 * 3 * sizeof(*d_box), cudaMemcpyDeviceToHost, stream));
+        // Verify that the cutoff and box size are valid together. If cutoff is greater than half the box
+        // then a particle can interact with multiple periodic copies.
+        const double db_cutoff = (cutoff_ + nblist_padding_) * 2;
+        cudaStreamSynchronize(stream);
+
+        // Verify that box is orthogonal and the width of the box in all dimensions is greater than twice the cutoff
+        for (int i = 0; i < 9; i++) {
+            if (i == 0 || i == 4 || i == 8) {
+                if (h_box[i] < db_cutoff) {
+                    throw std::runtime_error(
+                        "Cutoff with padding is more than half of the box width, neighborlist is no longer reliable");
+                }
+            } else if (h_box[i] != 0.0) {
+                throw std::runtime_error("Provided non-ortholinear box, unable to compute nonbonded energy");
+            }
+        }
+
+        gpuErrchk(cudaMemsetAsync(d_rebuild_nblist_, 0, sizeof(*d_rebuild_nblist_), stream));
+        gpuErrchk(cudaMemcpyAsync(d_nblist_x_, d_x, N * 3 * sizeof(*d_x), cudaMemcpyDeviceToDevice, stream));
+        gpuErrchk(cudaMemcpyAsync(d_nblist_box_, d_box, 3 * 3 * sizeof(*d_box), cudaMemcpyDeviceToDevice, stream));
+    } else {
+        k_permute<<<dimGrid, tpb, 0, stream>>>(N, d_perm_, d_x, d_sorted_x_);
+        gpuErrchk(cudaPeekAtLastError());
+    }
+
+    // do parameter interpolation here
+    if (Interpolated) {
+        CUresult result = compute_permute_interpolated_.configure(dimGrid, tpb, 0, stream)
+                              .launch(lambda, N, d_perm_, d_p, d_sorted_p_, d_sorted_dp_dl_);
+        if (result != 0) {
+            throw std::runtime_error("Driver call to k_permute_interpolated failed");
+        }
+    } else {
+        k_permute<<<dimGrid, tpb, 0, stream>>>(N, d_perm_, d_p, d_sorted_p_);
+        gpuErrchk(cudaPeekAtLastError());
+        gpuErrchk(cudaMemsetAsync(d_sorted_dp_dl_, 0, N * 3 * sizeof(*d_sorted_dp_dl_), stream))
+    }
+
+    // this stream needs to be synchronized so we can be sure that p_ixn_count_ is properly set.
```

Comment:
This comment seems to be no longer needed as there isn't a synchronization right here

Reply:
I moved the comment to where I think it should be, and also moved the call to `cudaStreamSynchronize` to right after the relevant copy in 43fd522. Can you see any downside to moving it above the box copy?